File: rtmppublisher/src/main/java/com/today/im/opus/OpusUtils.java
package com.today.im.opus;

import org.jetbrains.annotations.NotNull;

public class OpusUtils {

    private static OpusUtils instance;

    private OpusUtils() {
        System.loadLibrary("opus-jni");
    }

    public static OpusUtils instance() {
        if (instance == null) {
            synchronized (OpusUtils.class) {
                if (instance == null) {
                    instance = new OpusUtils();
                }
            }
        }

        return instance;
    }

    public native long initEncoder(int samplingRate, int numberOfChannels, int application);

    public native long initDecoder(int samplingRate, int numberOfChannels);

    public final native long createEncoder(int sampleRateInHz, int channelConfig, int complexity);

    public final native long createDecoder(int sampleRateInHz, int channelConfig, int complexity);

    public final native int encode(long handle, @NotNull short[] lin, int offset, @NotNull byte[] encoded);

    public final native int encodeByte(long handle, @NotNull byte[] lin, int offset, @NotNull byte[] encoded);

    public final native int decode(long handle, @NotNull byte[] encoded, @NotNull short[] lin);

    public final native int decodeShorts(long handle, byte[] in, short[] out, int frames);

    public final native void destroyEncoder(long handle);

    public final native void destroyDecoder(long handle);
}

File: rtmppublisher/src/main/java/com/today/im/opus/OpusUtils.kt
/*
 * Coding by Zhonghua. from 18-9-14 下午5:56.
 */

package com.today.im.opus

/**
 * JNI操作
 */
public class OpusUtils {
    init {
        System.loadLibrary("opus-jni")
    }

    companion object {
        private var opusUtils: OpusUtils? = null
        fun getInstant(): OpusUtils {
            if (opusUtils == null) {
                synchronized(OpusUtils::class.java) {
                    if (opusUtils == null) {
                        opusUtils = OpusUtils()
                    }
                }
            }
            return opusUtils!!
        }
    }

    external fun createEncoder(sampleRateInHz: Int, channelConfig: Int, complexity: Int): Long
    external fun createDecoder(sampleRateInHz: Int, channelConfig: Int): Long
    external fun encode(handle: Long, lin: ShortArray, offset: Int, encoded: ByteArray): Int
    external fun decode(handle: Long, encoded: ByteArray, lin: ShortArray): Int
    external fun destroyEncoder(handle: Long)
    external fun destroyDecoder(handle: Long)
}